File: src/main/java/org/kite9/diagram/common/elements/mapping/GeneratedLayoutConnection.java
package org.kite9.diagram.common.elements.mapping;

import org.kite9.diagram.common.HintMap;
import org.kite9.diagram.common.elements.AbstractBiDirectional;
import org.kite9.diagram.model.Connected;
import org.kite9.diagram.model.Connection;
import org.kite9.diagram.model.Container;
import org.kite9.diagram.model.DiagramElement;
import org.kite9.diagram.model.Label;
import org.kite9.diagram.model.Temporary;
import org.kite9.diagram.model.Terminator;
import org.kite9.diagram.model.position.Direction;
import org.kite9.diagram.model.position.RouteRenderingInformation;
import org.kite9.diagram.model.position.RouteRenderingInformationImpl;
import org.kite9.diagram.model.style.BoxShadow;
import org.kite9.diagram.visualization.planarization.mgt.BorderEdge;

/**
 * This connection is used with a {@link ContainerLayoutEdge} and is used to create a layout between
 * two elements of a container.  
 * 
 * Also, with {@link BorderEdge}, when two containers border each other.
 */
public class GeneratedLayoutConnection extends AbstractBiDirectional<Connected> implements Connection, Temporary {
	
	public GeneratedLayoutConnection(Connected from, Connected to, Direction drawDirection) {
		super(from, to, drawDirection);
	}

	@Override
	public String toString() {
		return "cle-"+getID();
	}

	@Override
	public int compareTo(DiagramElement o) {
		if (o instanceof AbstractBiDirectional<?>) {
			return this.getID().compareTo(((AbstractBiDirectional<?>) o).getID());
		} else {
			return -1;
		}
	}
	
	private RouteRenderingInformation rri;

	@Override
	public RouteRenderingInformation getRenderingInformation() {
		if (rri == null) {
			rri = new RouteRenderingInformationImpl();
		}
		
		return rri;
	}

	@Override
	public DiagramElement getParent() {
		return null;
	}
	
	public Container getContainer() {
		return null;
	}

	@Override
	public HintMap getPositioningHints() {
		return null;
	}

	@Override
	public int getDepth() {
		return 1;
	}

	@Override
	public Terminator getFromDecoration() {
		return null;
	}

	@Override
	public Terminator getToDecoration() {
		return null;
	}

	@Override
	public Label getFromLabel() {
		return null;
	}

	@Override
	public Label getToLabel() {
		return null;
	}

	@Override
	public int getRank() {
		return 0;
	}

	@Override
	public double getMargin(Direction d) {
		return 0;
	}

	@Override
	public double getPadding(Direction d) {
		return 0;
	}

	@Override
	public BoxShadow getShadow() {
		return null;
	}
}

File: src/main/kotlin/org/kite9/diagram/common/elements/mapping/GeneratedLayoutConnection.kt
package org.kite9.diagram.common.elements.mapping

import org.kite9.diagram.common.elements.AbstractBiDirectional
import org.kite9.diagram.model.*
import org.kite9.diagram.model.position.Direction
import org.kite9.diagram.model.position.Direction.Companion.reverse
import org.kite9.diagram.model.position.RouteRenderingInformation
import org.kite9.diagram.model.position.RouteRenderingInformationImpl

/**
 * This connection is used with a [ContainerLayoutEdge] and is used to create a layout between
 * two elements of a container.
 *
 * Also, with [BorderEdge], when two containers border each other.
 */
class GeneratedLayoutConnection(private val f: Connected, private val t: Connected, private val dd: Direction) :
    AbstractBiDirectional<Connected>(), Connection, Temporary {

    override fun toString(): String {
        return "glc-" + getID()
    }

    override operator fun compareTo(other: DiagramElement): Int {
        return if (other is AbstractBiDirectional<*>) {
            getID().compareTo((other as AbstractBiDirectional<*>).getID())
        } else {
            -1
        }
    }

    private val rri: RouteRenderingInformation = RouteRenderingInformationImpl()
    private val id = f.getID()+"-glc-"+t.getID();

    override fun getRenderingInformation(): RouteRenderingInformation {
        return rri
    }

    override fun getParent(): DiagramElement? {
        return null
    }

    val container: Container?
        get() = null

    override fun getDepth(): Int {
        return 1
    }

    override fun getFromDecoration(): Terminator? {
        return null
    }

    override fun getToDecoration(): Terminator? {
        return null
    }

    override fun getFromLabel(): Label? {
        return null
    }

    override fun getToLabel(): Label? {
        return null
    }

    override fun getRank(): Int {
        return 0
    }

    override fun getMargin(d: Direction): Double {
        return 0.0
    }

    override fun getPadding(d: Direction): Double {
        return 0.0
    }

    override fun getDecorationForEnd(end: DiagramElement): Terminator? {
        return null
    }

    override fun getMinimumLength(): Double {
        return 0.0
    }

    override fun getCornerRadius(): Double {
        return 0.0
    }

    override fun getFromArrivalSide(): Direction? {
        return reverse(dd)
    }

    override fun getToArrivalSide(): Direction? {
        return dd
    }

    override fun getFrom(): Connected {
        return f
    }

    override fun getTo(): Connected {
        return t
    }

    override fun getDrawDirection(): Direction? {
        return dd
    }

    override fun getID(): String {
        return id
    }


}